File: src/main/kotlin/net/starlegacy/feature/gas/collectionfactors/CollectionFactor.java
package net.starlegacy.feature.gas.collectionfactors;

import org.bukkit.Bukkit;
import org.bukkit.Location;

import java.util.ArrayList;
import java.util.List;

public abstract class CollectionFactor {

    public static List<CollectionFactor> collectionSetFromString(String string) {
        List<CollectionFactor> collectionFactors = new ArrayList<>();
        for (String text : string.split(";")) collectionFactors.add(valueOf(text));
        return collectionFactors;
    }

    private static CollectionFactor valueOf(String text) {
        String[] params = text.split(":");
        switch (params[0].toLowerCase()) {
            case "atmosphereheight":
                return new AtmosphereHeightFactor(Integer.parseInt(params[1]), Integer.parseInt(params[2]));
            case "distance":
                String[] locationParams = params[1].split(",");
                return new DistanceFactor(
                        new Location(Bukkit.getWorld(locationParams[0]), Integer.parseInt(locationParams[1]),
                                     Integer.parseInt(locationParams[2]), Integer.parseInt(locationParams[3])),
                        Integer.parseInt(params[2]), Float.parseFloat(params[3]));
            case "random":
                return new RandomFactor(Float.parseFloat(params[1]));
            case "hyperspaceonly":
                return new HyperspaceOnlyFactor();
            case "spaceonly":
                return new SpaceOnlyFactor();
            case "worldlimit":
                return new WorldLimitFactor(params[1].split(","));
            case "worldchance":
                return new WorldChanceFactor(Float.parseFloat(params[1]), params[2]);
            case "skylight":
            case "outdoors":
                return new OutdoorsFactor();
        }
        return new RandomFactor(1.0f);
    }

    public abstract boolean factor(Location location);
}

File: server/src/main/kotlin/net/horizonsend/ion/server/features/gas/collectionfactors/CollectionFactor.kt
package net.horizonsend.ion.server.features.gas.collectionfactors

import org.bukkit.Bukkit
import org.bukkit.Location
import java.util.Locale

abstract class CollectionFactor {
    abstract fun factor(location: Location): Boolean

	abstract fun canBeFound(location: Location): Boolean

    companion object {
        fun valueOf(text: String): CollectionFactor {
            val params = text.split(":".toRegex()).dropLastWhile { it.isEmpty() }

            when (params[0].lowercase(Locale.getDefault())) {
                "atmosphereheight" -> return AtmosphereHeightFactor(params[1].toInt().toDouble(), params[2].toInt().toDouble())
                "distance" -> {
                    val locationParams = params[1].split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    return DistanceFactor(
                        Location(
                            Bukkit.getWorld(locationParams[0]), locationParams[1].toInt()
                                .toDouble(), locationParams[2].toDouble(), locationParams[3].toDouble()
                        ), params[2].toInt().toDouble(), params[3].toFloat()
                    )
                }
                "random" -> return RandomFactor(params[1].toFloat())
                "hyperspaceonly" -> return HyperspaceOnlyFactor()
                "spaceonly" -> return SpaceOnlyFactor()
                "worldlimit" -> return WorldLimitFactor(params[1].split(",".toRegex()).dropLastWhile { it.isEmpty() })

                "worldchance" -> return WorldChanceFactor(
					params[1].split(",".toRegex()).dropLastWhile { it.isEmpty() }.associate {
						println(it)
						val pair = it.split("=".toRegex())

						pair[0] to pair[1].toDouble()
					}
				)

                "skylight", "outdoors" -> return OutdoorsFactor()
				"randomheight" -> return RandomByHeightFactor(params[1].toDouble(), params[2].toDouble(), params[3].toDouble(), params[4].toDouble())
            }

            return RandomFactor(1.0f)
        }
    }
}
